File: leetcode/16_med_3sum-closest.cpp
class Solution
{
public:
    int threeSumClosest(vector<int> &nums, int target)
    {
        int front, back, diff, res, min_res = INT_MAX;
        sort(nums.begin(), nums.end());
        for (int i = 0; i < nums.size(); i++)
        {
            front = i + 1;
            back = nums.size() - 1;
            while (front < back)
            {
                diff = target - nums[i] - nums[front] - nums[back];
                if (abs(diff) < min_res)
                {
                    min_res = abs(diff);
                    res = nums[i] + nums[front] + nums[back];
                }
                if (diff > 0)
                    front++;
                else if (diff < 0)
                    back--;
                else
                    return target;
            }
        }
        return res;
    }
};

File: leetcode/16_med_3sum-closest.py
class Solution(object):
    def threeSumClosest(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        nums.sort()
        return self.recursiveCloset(nums, target, 3)

    def recursiveCloset(self, nums, target, k):
        print("recursive", nums, target, k)
        n = len(nums)
        if n < k:
            return (10**4)+1000+1
        if n == k:
            return sum(nums[:k])
        
        # since nums is sorted
        if sum(nums[:k]) >= target:
            return sum(nums[:k])

        if sum(nums[-k:]) <= target:
            return sum(nums[-k:])

        # Find closest element
        if k == 1:
            difference = [(abs(target-x), x) for x in nums]
            return min(difference, key=lambda x: x[0])[1]

        closest = sum(nums[:k])
        for i, x in enumerate(nums):
            # If previous number is same, then computation has already occured
            if i > 0 and nums[i-1] == x: 
                continue 
            best_match = self.recursiveCloset(nums[i+1:], target-x, k-1)
            current_attempt = x + best_match
            if target == current_attempt:
                return current_attempt
            if abs(target - current_attempt) < abs(target-closest):
                closest = current_attempt
        return closest
